Improve parser detection of unhandled content

struan opened the pull request:
The parser now tracks all the tags it sees as it goes using tag IDs and then compares those to a list of IDs extracted using XPath. If there is a difference between the lists it throws an Exception.

There's also a number of parser improvements in here which were found in the process of making sure that it parsed things correctly:

* Fix for the parser failing to pick up all the text if there is more than one hs_Para element inside a Question tag
* Fixes broken table parsing code
* Fixes missing some content inside division tags
* Correctly handles clause tags to be part of the immediately following Amendment
* Makes hs_2cDebatedMotion a major heading
* Fixes missing some content inside new debate tags.

It also adds a script to make re-parsing easier.

Fixes #54
Fixes #66

struan committed:
handle mutiple para tags in a debate question

Fix for the parser failing to pick up all the text if there is more than
one hs_Para element instite a Question tag

dracos commented on pyscraper/new_hansard.py:
```diff
@@ -617,6 +617,15 @@ def parse_question(self, question):
         tag.append(p)
+
+        if len(para) > 1:
+            for p in para:
```
Does this not double the output in the cases it's trying to catch? e.g.
```
<Question><hs_Para><Number>Q2</Number>.
<Uin>[908984]</Uin>
<Member><B>Mr
Steve Reed</B> (Croydon North) (Lab):</Member>
<QuestionText></QuestionText>I
add my condolences to those already expressed about the former Father
of the House, and I welcome
my<hs_TimeCode time="2017-03-01T12:21:31"></hs_TimeCode> new hon.
Friend the Member for Stoke-on-Trent Central (Gareth Snell) to his
place.</hs_Para><hs_Para>Young
black men who use mental health services are more likely than other
people to be subject to detention, extreme forms of medication and
severe physical restraint, and, in extreme cases, this has led to
death, including that of my constituent Seni Lewis. Too many black
people with mental ill health are afraid to seek treatment from a
service they fear will not treat them fairly. Will the Prime Minister
meet me and some of the affected families to discuss the need for an
inquiry into institutional racism in the mental health
service<hs_TimeCode time="2017-03-01T12:22:18"></hs_TimeCode>?</hs_Para></Question>
```

The following-sibling would catch the first para after QuestionText, and then this loop would catch it again.

struan committed:
Keep a track of all the tag we've processed

Store the UID and HRSContentID of handled tags so we can later compare
to a list of all IDs in the document

struan committed:
throw an exception if it looks like we missed a tag

Get a list of all tag IDs in the document and compare to the list we've
processed and throw an exception if they don't match.

struan committed:
correctly parse tables

We didn't use namespaces before so they weren't being parsed properly.
Correct this and track the tags.

struan committed:
improve question parsing and track the tags

Make sure we are coping with questions where part of the question isn't
in the tail of QuestionText but is in following tags.

Also cope with oddities like multiple question number tags.

struan committed:
handle clause tags in the following heading

Clause tags actually relate to the text after them so ignore them at the
top level and then go back and parse them as part of the following
heading tag. Then add them as the first part of the first speech under
the heading.

Fixes #53

struan committed:
cope with multiple debate heading and procedure tags in new debate

If there's more than one heading or procedure in a new debate tag then
make those into paragraphs in the first speech of the debate.

struan committed:
parse time tags inside a division tag

struan committed:
gather up all the text inside a division tag

struan committed:
hs_2cDebatedMotion is actually a major heading

struan committed:
better parsing for Lords Amemdments

rather than just parsing it all into a single line of text parse all the
paragraphs and indents so that we try and retain a bit more structure.

struan committed:
Script to reparse all hansard zip file contents

Scans the list of seen files and then picks out the latest one
and then re-parses that. Assumes that the files are ordered in
date order in the list.

struan committed:
mark as seen tags not always accessed via handle_tag

Copes with tags that are mostly processed from inside another tag

struan committed:
parse english and welsh only division counts

Fixes #63

struan committed:
mark all the paragraph and time tags in a division as seen

struan committed:
mark various bits of structure as seen

There's lots of tags that we don't directly parse as we're interested in
sub tags or they are parsed as part of the parent. Mark these as seen.

struan committed:
handle times with a newline between hours and minutes
